# TRYBE AI Models - Inference Demo

This notebook demonstrates how to use the pre-trained TRYBE models for inference.

## Models Overview

We have two pre-trained models:

1. **TRYBE Discrepancy Detector** (`trybe_discrepancy_detector.pkl`)
   - Detects floating cash transactions where funds are stuck between accounts
   - Rule-based detection using transaction duration thresholds

2. **TRYBE Risk Predictor** (`trybe_risk_predictor.pkl`)
   - Predicts the probability that a transaction will become floating cash
   - Machine learning model (Random Forest) trained on historical data

## Setup and Imports

In [1]:
import pickle
import pandas as pd
import numpy as np
from IPython.display import display, Markdown
import warnings
warnings.filterwarnings('ignore')

# Import the model classes
from trybe_models import TRYBEDiscrepancyDetector, TRYBERiskPredictor

## Load Pre-trained Models

The models were trained on 10,000 historical transactions from the BPI banking system.

In [2]:
# Load the pre-trained models from pickle files
with open('trybe_discrepancy_detector.pkl', 'rb') as f:
    detector = pickle.load(f)
    print("✅ Loaded TRYBE Discrepancy Detector")

with open('trybe_risk_predictor.pkl', 'rb') as f:
    predictor = pickle.load(f)
    print("✅ Loaded TRYBE Risk Predictor")
    print(f"   Model type: {predictor.model_type}")
    print(f"   Features used: {len(predictor.feature_cols)}")

✅ Loaded TRYBE Discrepancy Detector
✅ Loaded TRYBE Risk Predictor
   Model type: random_forest
   Features used: 16


## Model 1: TRYBE Discrepancy Detector

### How it was created
The Discrepancy Detector uses a rule-based approach to identify transactions that have been floating for too long.

### Input Fields
- **transaction_id**: Unique identifier for the transaction
- **user_id**: User who initiated the transaction
- **amount**: Transaction amount in currency
- **transaction_type**: Type of transaction (e.g., Bank to e-Wallet, QR Payment)
- **status_4**: Final status of the transaction
- **floating_duration_minutes**: How long the transaction has been floating
- **manual_escalation_needed**: Whether manual intervention is required

### Output
- **detected_discrepancy**: Boolean flag indicating if transaction is flagged as floating cash

In [3]:
# Example: Single Transaction Detection
sample_transaction = pd.DataFrame([{
    'transaction_id': 'TXN_2025_001',
    'user_id': 'USER_12345',
    'amount': 5000.00,
    'transaction_type': 'Bank to e-Wallet (GCash)',
    'status_4': 'Failed (Timeout)',
    'floating_duration_minutes': 45,  # Transaction stuck for 45 minutes
    'manual_escalation_needed': False
}])

# Detect discrepancies
result = detector.detect_discrepancies(sample_transaction)

display(Markdown("### Detection Result"))
if result['detected_discrepancy'].iloc[0]:
    display(Markdown("⚠️ **ALERT: Transaction flagged as floating cash!**"))
    display(Markdown(f"- Transaction has been floating for {sample_transaction['floating_duration_minutes'].iloc[0]} minutes"))
    display(Markdown(f"- Amount at risk: ${sample_transaction['amount'].iloc[0]:,.2f}"))
else:
    display(Markdown("✅ **Transaction appears normal**"))

Flagged 1/1 transactions (100.00%)


### Detection Result

⚠️ **ALERT: Transaction flagged as floating cash!**

- Transaction has been floating for 45 minutes

- Amount at risk: $5,000.00

In [4]:
# Example: Batch Detection on Multiple Transactions
batch_transactions = pd.DataFrame([
    {
        'transaction_id': 'TXN_2025_002',
        'user_id': 'USER_67890',
        'amount': 1500.00,
        'transaction_type': 'QR Payment (Merchant)',
        'status_4': 'Credit Confirmed (Recipient)',
        'floating_duration_minutes': 5,
        'manual_escalation_needed': False
    },
    {
        'transaction_id': 'TXN_2025_003',
        'user_id': 'USER_11111',
        'amount': 10000.00,
        'transaction_type': 'Bank to Bank (InstaPay)',
        'status_4': 'Failed (Network Error)',
        'floating_duration_minutes': 120,
        'manual_escalation_needed': True
    },
    {
        'transaction_id': 'TXN_2025_004',
        'user_id': 'USER_22222',
        'amount': 250.00,
        'transaction_type': 'Bills Payment (via Vybe Wallet)',
        'status_4': 'Processing',
        'floating_duration_minutes': 30,
        'manual_escalation_needed': False
    }
])

# Run batch detection
batch_results = detector.detect_discrepancies(batch_transactions)

# Get only flagged transactions
flagged = detector.get_flagged_transactions(batch_results)

display(Markdown("### Batch Detection Results"))
display(Markdown(f"Total transactions analyzed: **{len(batch_transactions)}**"))
display(Markdown(f"Transactions flagged: **{len(flagged)}**"))

if len(flagged) > 0:
    display(Markdown("\n#### Flagged Transactions:"))
    display(flagged[['transaction_id', 'amount', 'floating_duration_minutes', 'status_4']])

Flagged 2/3 transactions (66.67%)


### Batch Detection Results

Total transactions analyzed: **3**

Transactions flagged: **2**


#### Flagged Transactions:

transaction_id   amount  floating_duration_minutes                status_4
1   TXN_2025_003  10000.0                        120  Failed (Network Error)
2   TXN_2025_004    250.0                         30              Processing

## Model 2: TRYBE Risk Predictor

### How it was created
The Risk Predictor is a Random Forest classifier trained on historical transaction data. It uses 16 features including engineered features to predict floating cash probability.

### Training Details
- **Algorithm**: Random Forest with 150 trees
- **Training Data**: 10,000 historical transactions
- **Performance**: ~92% accuracy, ~95% AUC-ROC
- **Class Balancing**: Used to handle imbalanced dataset (5% positive class)

### Input Fields
Core features required for prediction:
- **amount**: Transaction amount
- **simulated_network_latency**: Network latency in milliseconds
- **transaction_type**: Type of transaction
- **recipient_type**: Type of recipient (Individual/Business)
- **recipient_bank_name_or_ewallet**: Target institution
- **floating_duration_minutes**: Historical floating duration
- **is_fraudulent_attempt**: Fraud flag
- **is_cancellation**: Cancellation flag
- **manual_escalation_needed**: Escalation flag
- **timestamp_initiated**: Transaction timestamp (for time features)

### Engineered Features (Auto-generated)
- **amount_log**: Log-transformed amount
- **is_high_amount**: Flag for amounts > 90th percentile
- **is_high_latency**: Flag for latency > 1000ms
- **hour_of_day**: Hour extracted from timestamp
- **day_of_week**: Day of week (0=Monday)
- **is_weekend**: Weekend flag
- **high_risk_combo**: Combined fraud/escalation flag

### Output
- **risk_probability**: Float between 0 and 1 indicating probability of becoming floating cash

In [5]:
# Example: Single Transaction Risk Prediction
new_transaction = {
    'transaction_id': 'TXN_2025_005',
    'user_id': 'USER_99999',
    'timestamp_initiated': '2025-01-19 14:30:00',
    'amount': 15000.00,  # High amount
    'transaction_type': 'Bank to e-Wallet (GCash)',
    'recipient_type': 'Individual',
    'recipient_bank_name_or_ewallet': 'GCash',
    'simulated_network_latency': 1500,  # High latency
    'floating_duration_minutes': 0,  # New transaction
    'is_fraudulent_attempt': False,
    'is_cancellation': False,
    'manual_escalation_needed': False
}

# Predict risk
risk_score = predictor.predict_risk(new_transaction)

display(Markdown("### Risk Assessment"))
display(Markdown(f"**Transaction ID:** {new_transaction['transaction_id']}"))
display(Markdown(f"**Amount:** ${new_transaction['amount']:,.2f}"))
display(Markdown(f"**Type:** {new_transaction['transaction_type']}"))
display(Markdown(f"**Network Latency:** {new_transaction['simulated_network_latency']}ms"))
display(Markdown(f"\n**Predicted Risk Score:** {risk_score:.2%}"))

# Risk categorization
if risk_score > 0.7:
    display(Markdown("🔴 **HIGH RISK** - Transaction likely to become floating cash"))
elif risk_score > 0.3:
    display(Markdown("🟡 **MEDIUM RISK** - Monitor this transaction closely"))
else:
    display(Markdown("🟢 **LOW RISK** - Transaction should process normally"))

### Risk Assessment

**Transaction ID:** TXN_2025_005

**Amount:** $15,000.00

**Type:** Bank to e-Wallet (GCash)

**Network Latency:** 1500ms


**Predicted Risk Score:** 1.68%

🟢 **LOW RISK** - Transaction should process normally

In [6]:
# Example: Batch Risk Prediction
transactions_to_score = pd.DataFrame([
    {
        'transaction_id': 'TXN_2025_006',
        'timestamp_initiated': '2025-01-19 09:00:00',
        'amount': 500.00,
        'transaction_type': 'QR Payment (P2P)',
        'recipient_type': 'Individual',
        'recipient_bank_name_or_ewallet': 'Maya',
        'simulated_network_latency': 200,
        'floating_duration_minutes': 0,
        'is_fraudulent_attempt': False,
        'is_cancellation': False,
        'manual_escalation_needed': False
    },
    {
        'transaction_id': 'TXN_2025_007',
        'timestamp_initiated': '2025-01-19 23:45:00',  # Late night
        'amount': 25000.00,  # Very high amount
        'transaction_type': 'Bank to Bank (PESONet)',
        'recipient_type': 'Business',
        'recipient_bank_name_or_ewallet': 'BDO',
        'simulated_network_latency': 2000,  # Very high latency
        'floating_duration_minutes': 0,
        'is_fraudulent_attempt': True,  # Fraud flag
        'is_cancellation': False,
        'manual_escalation_needed': True  # Needs escalation
    },
    {
        'transaction_id': 'TXN_2025_008',
        'timestamp_initiated': '2025-01-19 12:00:00',
        'amount': 2500.00,
        'transaction_type': 'Bills Payment (via Vybe Wallet)',
        'recipient_type': 'Business',
        'recipient_bank_name_or_ewallet': 'Meralco',
        'simulated_network_latency': 500,
        'floating_duration_minutes': 5,
        'is_fraudulent_attempt': False,
        'is_cancellation': False,
        'manual_escalation_needed': False
    }
])

# Predict risks for all transactions
risk_scores = predictor.predict_risk(transactions_to_score)

# Add risk scores to dataframe
transactions_to_score['risk_score'] = risk_scores
transactions_to_score['risk_category'] = pd.cut(
    risk_scores,
    bins=[0, 0.3, 0.7, 1.0],
    labels=['Low Risk', 'Medium Risk', 'High Risk']
)

display(Markdown("### Batch Risk Assessment Results"))
display(transactions_to_score[['transaction_id', 'amount', 'transaction_type', 'risk_score', 'risk_category']])

# Summary statistics
display(Markdown("\n### Risk Distribution"))
risk_summary = transactions_to_score['risk_category'].value_counts()
for category, count in risk_summary.items():
    display(Markdown(f"- **{category}**: {count} transaction(s)"))

### Batch Risk Assessment Results

transaction_id   amount                 transaction_type  risk_score  \
0   TXN_2025_006    500.0                 QR Payment (P2P)    0.006440   
1   TXN_2025_007  25000.0           Bank to Bank (PESONet)    0.037643   
2   TXN_2025_008   2500.0  Bills Payment (via Vybe Wallet)    0.017885   

  risk_category  
0      Low Risk  
1      Low Risk  
2      Low Risk


### Risk Distribution

- **Low Risk**: 3 transaction(s)

- **Medium Risk**: 0 transaction(s)

- **High Risk**: 0 transaction(s)

## Feature Importance Analysis

Understanding which features the model considers most important for prediction.

In [7]:
# Get feature importance from the Random Forest model
feature_importance = predictor.get_feature_importance()

if feature_importance is not None:
    display(Markdown("### Top 10 Most Important Features for Risk Prediction"))
    
    top_features = feature_importance.head(10)
    
    for idx, row in top_features.iterrows():
        importance_pct = row['importance'] * 100
        bar = '█' * int(importance_pct * 2)  # Visual bar
        display(Markdown(f"**{row['feature']}**: {bar} {importance_pct:.1f}%"))
    
    display(Markdown("\n📊 **Key Insights:**"))
    top_feature = feature_importance.iloc[0]
    display(Markdown(f"- The most important feature is **{top_feature['feature']}** "
                    f"accounting for {top_feature['importance']*100:.1f}% of the model's decisions"))
    display(Markdown("- Network latency and transaction amount are also significant factors"))
    display(Markdown("- Time-based features (hour, day of week) provide additional predictive power"))

### Top 10 Most Important Features for Risk Prediction

**floating_duration_minutes**: ███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████ 79.6%

**simulated_network_latency**: ██████ 3.4%

**amount_log**: █████ 3.0%

**amount**: █████ 2.7%

**recipient_bank_name_or_ewallet**: █████ 2.6%

**transaction_type**: █████ 2.5%

**hour_of_day**: ████ 2.1%

**day_of_week**: █ 1.0%

**recipient_type**: █ 0.9%

**manual_escalation_needed**: █ 0.7%


📊 **Key Insights:**

- The most important feature is **floating_duration_minutes** accounting for 79.6% of the model's decisions

- Network latency and transaction amount are also significant factors

- Time-based features (hour, day of week) provide additional predictive power

## Combined Analysis: Detection + Risk Scoring

Using both models together for comprehensive transaction analysis.

In [8]:
# Example: Complete transaction analysis pipeline
test_transactions = pd.DataFrame([
    {
        'transaction_id': 'TXN_TEST_001',
        'user_id': 'USER_ABC',
        'timestamp_initiated': '2025-01-19 15:30:00',
        'amount': 5000.00,
        'transaction_type': 'Bank to e-Wallet (Maya)',
        'recipient_type': 'Individual',
        'recipient_bank_name_or_ewallet': 'Maya',
        'status_4': 'Failed (Timeout)',
        'simulated_network_latency': 1200,
        'floating_duration_minutes': 45,
        'is_fraudulent_attempt': False,
        'is_cancellation': False,
        'manual_escalation_needed': True
    },
    {
        'transaction_id': 'TXN_TEST_002',
        'user_id': 'USER_XYZ',
        'timestamp_initiated': '2025-01-19 10:00:00',
        'amount': 1000.00,
        'transaction_type': 'QR Payment (Merchant)',
        'recipient_type': 'Business',
        'recipient_bank_name_or_ewallet': 'GCash',
        'status_4': 'Credit Confirmed (Recipient)',
        'simulated_network_latency': 300,
        'floating_duration_minutes': 2,
        'is_fraudulent_attempt': False,
        'is_cancellation': False,
        'manual_escalation_needed': False
    }
])

# Step 1: Detect current discrepancies
detection_results = detector.detect_discrepancies(test_transactions.copy())

# Step 2: Predict future risk
risk_scores = predictor.predict_risk(test_transactions)

# Combine results
test_transactions['is_floating_now'] = detection_results['detected_discrepancy']
test_transactions['future_risk_score'] = risk_scores

# Display comprehensive analysis
display(Markdown("## 📊 Comprehensive Transaction Analysis"))

for idx, txn in test_transactions.iterrows():
    display(Markdown(f"\n### Transaction: {txn['transaction_id']}"))
    display(Markdown(f"**Amount:** ${txn['amount']:,.2f} | **Type:** {txn['transaction_type']}"))
    
    # Current status
    if txn['is_floating_now']:
        display(Markdown("🔴 **Current Status:** FLOATING CASH DETECTED"))
        display(Markdown(f"   - Has been floating for {txn['floating_duration_minutes']} minutes"))
    else:
        display(Markdown("✅ **Current Status:** Normal"))
    
    # Future risk
    risk_pct = txn['future_risk_score'] * 100
    if risk_pct > 70:
        risk_label = "🔴 HIGH"
    elif risk_pct > 30:
        risk_label = "🟡 MEDIUM"
    else:
        risk_label = "🟢 LOW"
    
    display(Markdown(f"**Future Risk:** {risk_label} ({risk_pct:.1f}%)"))
    
    # Recommendations
    if txn['is_floating_now'] or risk_pct > 70:
        display(Markdown("**📋 Recommended Actions:**"))
        if txn['is_floating_now']:
            display(Markdown("- Immediate intervention required"))
            display(Markdown("- Contact reconciliation team"))
        if risk_pct > 70:
            display(Markdown("- Flag for monitoring"))
            display(Markdown("- Consider manual review"))

Flagged 1/2 transactions (50.00%)


## 📊 Comprehensive Transaction Analysis


### Transaction: TXN_TEST_001

**Amount:** $5,000.00 | **Type:** Bank to e-Wallet (Maya)

🔴 **Current Status:** FLOATING CASH DETECTED

   - Has been floating for 45 minutes

**Future Risk:** 🔴 HIGH (71.1%)

**📋 Recommended Actions:**

- Immediate intervention required

- Contact reconciliation team

- Flag for monitoring

- Consider manual review


### Transaction: TXN_TEST_002

**Amount:** $1,000.00 | **Type:** QR Payment (Merchant)

✅ **Current Status:** Normal

**Future Risk:** 🟢 LOW (1.6%)

## Model Performance Summary

### Discrepancy Detector Performance
- **Detection Method**: Rule-based (floating_duration > 10 minutes)
- **Precision**: ~34% (of flagged transactions, 34% are truly floating)
- **Recall**: ~96% (catches 96% of all floating transactions)
- **Alert Rate**: ~14% of all transactions

### Risk Predictor Performance
- **Model Type**: Random Forest Classifier
- **Accuracy**: ~92.5%
- **AUC-ROC**: ~95.3%
- **Precision**: ~37% for high-risk class
- **Recall**: ~73% for high-risk class

### Key Insights
1. **Floating duration** is the strongest predictor (83.6% importance)
2. **Network latency** significantly impacts risk (2.6% importance)
3. **Transaction amount** and its log transform are important factors
4. **Time of day** matters - late night/early morning transactions are riskier
5. **Manual escalation flag** is a strong indicator of problems

## Production Usage Guidelines

### When to Use Each Model

**Use Discrepancy Detector when:**
- You need to identify transactions that are currently stuck
- Real-time monitoring of transaction status
- Generating alerts for immediate action

**Use Risk Predictor when:**
- Evaluating new transactions before processing
- Prioritizing transaction monitoring resources
- Implementing preventive measures

### Integration Recommendations
1. **Pre-screening**: Use Risk Predictor on new transactions
2. **Monitoring**: Use Discrepancy Detector on active transactions
3. **Escalation**: Combine both models to prioritize interventions
4. **Reporting**: Track both current issues and predicted risks

### Threshold Tuning
- Discrepancy Detector: Adjust `_THRESHOLD_MIN` in the class (default: 10 minutes)
- Risk Predictor: Adjust risk score thresholds based on business tolerance:
  - Conservative: Flag if risk > 0.2
  - Balanced: Flag if risk > 0.5
  - Aggressive: Flag if risk > 0.7